ArchiSteamFarm/Statistics.cs

```diff
@@ -457,22 +463,64 @@ internal sealed class Statistics : IDisposable {
 						do {
 							match = false;
 
-							foreach ((ulong ourItem, uint ourAmount) in ourFullSet.Where(item => item.Value > 1).OrderByDescending(item => item.Value)) {
-								if (!ourTradableSet.TryGetValue(ourItem, out uint tradableAmount) || (tradableAmount == 0)) {
+							foreach ((ulong ourItem, uint ourFullAmount) in ourFullSet.Where(item => item.Value > 1).OrderByDescending(item => item.Value)) {
+								if (!ourTradableSet.TryGetValue(ourItem, out uint ourTradableAmount) || (ourTradableAmount == 0)) {
 									continue;
 								}
 
-								foreach ((ulong theirItem, _) in theirItems.OrderBy(item => ourFullSet.TryGetValue(item.Key, out uint ourAmountOfTheirItem) ? ourAmountOfTheirItem : 0)) {
-									if (ourFullSet.TryGetValue(theirItem, out uint ourAmountOfTheirItem) && (ourAmount <= ourAmountOfTheirItem + 1)) {
+								foreach ((ulong theirItem, uint theirFullAmount) in theirFullItems.OrderBy(item => ourFullSet.TryGetValue(item.Key, out uint ourAmountOfTheirItem) ? ourAmountOfTheirItem : 0)) {
+									if (!theirTradableItems.TryGetValue(theirItem, out uint theirTradableAmount) || (theirTradableAmount == 0)) {
+										continue;
+									}
+
+									if (ourFullSet.TryGetValue(theirItem, out uint ourAmountOfTheirItem) && (ourFullAmount <= ourAmountOfTheirItem + 1)) {
+										continue;
+									}
+
+									if (theirFullItems.TryGetValue(ourItem, out uint theirAmountOfOurItem) && (theirFullAmount <= theirAmountOfOurItem + 1)) {
 										continue;
 									}
 
+									if (!listedUser.MatchEverything) {
+										// We have a potential match, let's check fairness for them
+										fairClassIDsToGive[ourItem] = fairClassIDsToGive.TryGetValue(ourItem, out uint fairGivenAmount) ? fairGivenAmount + 1 : 1;
+										fairClassIDsToReceive[theirItem] = fairClassIDsToReceive.TryGetValue(theirItem, out uint fairReceivedAmount) ? fairReceivedAmount + 1 : 1;
+
+										// Convert list to HashSet<Steam.Asset>
+										HashSet<Steam.Asset> fairItemsToGive = Trading.GetTradableItemsFromInventory(ourInventory, fairClassIDsToGive);
+										HashSet<Steam.Asset> fairItemsToReceive = Trading.GetTradableItemsFromInventory(theirInventory, fairClassIDsToReceive);
+
+										// Filter inventory for the sets we're looking for for IsTradeNeutralOrBetter
+										HashSet<Steam.Asset> fairFiltered = new HashSet<Steam.Asset>();
+
+										foreach (Steam.Asset item in theirInventory.Where(item => (item.RealAppID == set.RealAppID) && (item.Type == set.Type) && (item.Rarity == set.Rarity))) {
+											fairFiltered.Add(item.CreateShallowCopy());
+										}
+
+										// Actual check:
+										if (!Trading.IsTradeNeutralOrBetter(fairFiltered, fairItemsToReceive, fairItemsToGive)) {
+											if (fairGivenAmount > 1) {
```

Comment:
This `if` and the one below are entirely wrong since you're declaring `fairGivenAmount` as existing amount then adding 1 but not storing it in the variable, so `fairGivenAmount` no longer represents the **current** state of the dictionary at this moment.

Instead of fixing `fairGivenAmount` above, it's probably a better idea to just `TryGetValue(ourItem, out fairGivenAmount)` of up-to-date state.

Reply:
Alternatively you can fix `fairGivenAmount` above to reflect the current state:

```c#
fairClassIDsToGive.TryGetValue(ourItem, out uint fairGivenAmount);
fairClassIDsToGive[ourItem] = ++fairGivenAmount;
```

After second thought I think that it'll actually be better than another `if` block.

Reply:
That's a good one! Completely oversight that I haven't saved it again.